spec/std/system/user_spec.cr

```diff
@@ -0,0 +1,97 @@
+require "spec"
+require "system/user"
+
+private USER_NAME = "root"
+private USER_ID   = 0
+
+private GROUP_GID = 0
+{% if flag?(:darwin) %}
+  private USER_HOME = "/var/root"
+  private USER_SHELL = "/bin/sh"
+  private USER_INFO = "System Administrator"
+{% elsif flag?(:openbsd) %}
+  private USER_HOME = "/root"
+  private USER_SHELL = "/bin/ksh"
+  private USER_INFO = "Charlie &"
+{% elsif flag?(:freebsd) %}
+  private USER_HOME = "/root"
+  private USER_SHELL = "/bin/csh"
+  private USER_INFO = "Charlie &"
+{% elsif flag?(:linux) %}
+  private USER_HOME = "/root"
+  private USER_SHELL = "/bin/bash"
+  private USER_INFO = "root"
+{% else %}
+  {{ raise "Unsupported platform, only Darwin, OpenBSD, FreeBSD, and Linux are supported." }}
+{% end %}
+
+private BAD_USER_NAME = "non_existant_user"
+private BAD_USER_ID   = 1234567
+
+describe System::User do
+  it "username from uid" do
+    System::User.name(USER_ID).should eq(USER_NAME)
+    System::User.name?(USER_ID).should eq(USER_NAME)
+
+    expect_raises System::User::NotFoundError do
+      System::User.name(BAD_USER_ID)
+    end
+    System::User.name?(BAD_USER_ID).should be_nil
+  end
+
+  it "uid from username" do
+    System::User.uid(USER_NAME).should eq(USER_ID)
+    System::User.uid?(USER_NAME).should eq(USER_ID)
+
+    expect_raises System::User::NotFoundError do
+      System::User.uid(BAD_USER_NAME)
+    end
+    System::User.uid?(BAD_USER_NAME).should be_nil
+  end
+
+  it "gets a user" do
+    System::User.get(USER_NAME)
+    System::User.get(USER_ID)
+
+    System::User.get?(USER_NAME).should_not be_nil
+    System::User.get?(USER_ID).should_not be_nil
+  end
+
+  it "raises on user not found" do
+    expect_raises System::User::NotFoundError do
+      System::User.get(BAD_USER_NAME)
+    end
+    expect_raises System::User::NotFoundError do
+      System::User.get(BAD_USER_ID)
+    end
+
+    System::User.get?(BAD_USER_NAME).should be_nil
+    System::User.get?(BAD_USER_ID).should be_nil
+  end
+
+  it "raises UID out of bounds" do
+    System::User.check_uid_in_bounds(System::User::UID_MAX)
+
+    expect_raises System::User::OutOfBoundsError do
+      System::User.get(System::User::UID_MAX.to_u64 + 1)
+    end
+  end
+
+  it "has the correct properties" do
+    user = System::User.get(USER_NAME)
+    user.name.should eq(USER_NAME)
+    user.uid.should eq(USER_ID)
+    user.gid.should eq(GROUP_GID)
+    user.home.should eq(USER_HOME)
+    user.shell.should eq(USER_SHELL)
+    user.info.should eq(USER_INFO)
+
+    user = System::User.get(USER_ID)
```

Comment:
Just assert that all the info is the same between getting by name and by id. If we're reading garbage data, it's likely uninitialized memory so different between the two calls and will fail the spec.

Reply:
This isn't really necessary. All the info always comes from the same call (by `uid`/`gid`). A request by name determines the `uid`/`gid` of that name and then requests the data for that `uid`/`gid`.